glean-core/src/scheduler.rs

```diff
@@ -0,0 +1,515 @@
+        lifetime: Lifetime::User,
+        ..Default::default()
+    })
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::tests::new_glean;
+    use std::sync::atomic::{AtomicU32, Ordering};
+
+    struct ValidatingSubmitter<F: Fn(DateTime<FixedOffset>, Option<&str>)> {
+        submit_validator: F,
+        validator_run_count: Arc<AtomicU32>,
+    }
+    struct ValidatingScheduler<F: Fn(DateTime<FixedOffset>, When)> {
+        schedule_validator: F,
+        validator_run_count: Arc<AtomicU32>,
+    }
+    impl<F: Fn(DateTime<FixedOffset>, Option<&str>)> MetricsPingSubmitter for ValidatingSubmitter<F> {
+        fn submit_metrics_ping(
+            &self,
+            _glean: &Glean,
+            reason: Option<&str>,
+            now: DateTime<FixedOffset>,
+        ) {
+            (self.submit_validator)(now, reason);
+            self.validator_run_count.fetch_add(1, Ordering::Relaxed);
+        }
+    }
+    impl<F: Fn(DateTime<FixedOffset>, When)> MetricsPingScheduler for ValidatingScheduler<F> {
+        fn start_scheduler(
+            &self,
+            _submitter: impl MetricsPingSubmitter + Send + 'static,
+            now: DateTime<FixedOffset>,
+            when: When,
+        ) {
+            (self.schedule_validator)(now, when);
+            self.validator_run_count.fetch_add(1, Ordering::Relaxed);
+        }
+    }
+
+    fn new_proxies<
+        F1: Fn(DateTime<FixedOffset>, Option<&str>),
+        F2: Fn(DateTime<FixedOffset>, When),
+    >(
+        submit_validator: F1,
+        schedule_validator: F2,
+    ) -> (
+        ValidatingSubmitter<F1>,
+        Arc<AtomicU32>,
+        ValidatingScheduler<F2>,
+        Arc<AtomicU32>,
+    ) {
+        let submitter_count = Arc::new(AtomicU32::new(0));
+        let submitter = ValidatingSubmitter {
+            submit_validator,
+            validator_run_count: Arc::clone(&submitter_count),
+        };
+        let scheduler_count = Arc::new(AtomicU32::new(0));
+        let scheduler = ValidatingScheduler {
+            schedule_validator,
+            validator_run_count: Arc::clone(&scheduler_count),
+        };
+        (submitter, submitter_count, scheduler, scheduler_count)
+    }
+
+    // Ensure that if we have a different build, we immediately submit an "upgrade" ping
+    // and schedule a "reschedule" ping for tomorrow.
+    #[test]
+    fn different_app_builds_submit_and_reschedule() {
+        let (mut glean, _t) = new_glean(None);
+
+        glean.app_build = "a build".into();
+        get_last_sent_build_metric().set(&glean, "a different build");
+
+        let (submitter, submitter_count, scheduler, scheduler_count) = new_proxies(
+            |_, reason| assert_eq!(reason, Some("upgrade")),
+            |_, when| assert_eq!(when, When::Reschedule),
+        );
+
+        schedule_internal(&glean, submitter, scheduler, local_now_with_offset().0);
+        assert_eq!(1, submitter_count.swap(0, Ordering::Relaxed));
```

Comment:
Wondered why you use `swap(0)` here, when `load()` would work just as fine? (It's not wrong and can stay, just trying to understand your thinking)

Reply:
I use `swap(0)` so that if someone extends the test the `submitter_count` has been reset.